server/src/main/java/org/elasticsearch/search/vectors/KnnScoreDocQueryBuilder.java

```diff
@@ -100,14 +103,18 @@ protected QueryBuilder doRewrite(QueryRewriteContext queryRewriteContext) throws
         return super.doRewrite(queryRewriteContext);
     }
 
-    private static int[] findSegmentStarts(IndexReader reader, int[] docs) {
+    private static int[] findSegmentStarts(IndexReader reader, int[] docs, Object[] segmentIds) {
         int[] starts = new int[reader.leaves().size() + 1];
         starts[starts.length - 1] = docs.length;
+        LeafReader lastLeaf = reader.leaves().get(reader.leaves().size() - 1).reader();
+        lastLeaf = FilterLeafReader.unwrap(lastLeaf);
+        segmentIds[reader.leaves().size() - 1] = lastLeaf.getContext().id();
```

Comment:
From what I can tell, `FilterLeafReader` can have different context ids. This fails as we may be wrapped or not, or the internal thing will be unwrapped and wrapped a different way later :/